blockchain/message_pool/src/msgpool.rs

```diff
@@ -0,0 +1,866 @@
+/// future TODO is to add a pubsub field to allow for publishing updates.
+pub struct MpoolProvider<DB> {
+    cs: ChainStore<DB>,
+}
+
+impl<'db, DB> MpoolProvider<DB>
+where
+    DB: BlockStore,
+{
+    pub fn new(cs: ChainStore<DB>) -> Self
+    where
+        DB: BlockStore,
+    {
+        MpoolProvider { cs }
+    }
+}
+
+impl<DB> Provider for MpoolProvider<DB>
+where
+    DB: BlockStore,
+{
+    /// TODO find a way to get this function to periodically check for head changes in cs and update
+    /// msgpool accordingly or find a way to have an async function always running in the background
+    /// checking for newly published tipset and add it to msgpool
+    fn subscribe_head_changes(&mut self) -> Tipset {
+        if self.cs.heaviest_tipset().is_none() {
+            task::block_on(async {
+                let mut subscriber = self.cs.subscribe();
+                loop {
+                    let sub = subscriber.next().await;
+                    if sub.is_some()
+                        && self
+                            .cs
+                            .set_heaviest_tipset(sub.unwrap().clone())
+                            .await
+                            .is_ok()
+                    {
+                        break;
+                    }
+                }
+            });
+        }
+        self.cs.heaviest_tipset().unwrap().as_ref().clone()
+    }
+
+    fn put_message(&self, msg: &SignedMessage) -> Result<Cid, Error> {
+        let cid = self
+            .cs
+            .db
+            .put(msg, Blake2b256)
+            .map_err(|err| Error::Other(err.to_string()))?;
+        Ok(cid)
+    }
+
+    fn state_get_actor(&self, addr: &Address, ts: &Tipset) -> Result<ActorState, Error> {
+        let state = StateTree::new_from_root(self.cs.db.as_ref(), ts.parent_state())
+            .map_err(Error::Other)?;
+        let actor = state.get_actor(addr).map_err(Error::Other)?;
+        match actor {
+            Some(actor_state) => Ok(actor_state),
+            None => Err(Error::Other("No actor state".to_string())),
+        }
+    }
+
+    fn messages_for_block(
+        &self,
+        h: &BlockHeader,
+    ) -> Result<(Vec<UnsignedMessage>, Vec<SignedMessage>), Error> {
+        chain::block_messages(self.cs.blockstore(), h).map_err(|err| Error::Other(err.to_string()))
+    }
+
+    fn messages_for_tipset(&self, h: &Tipset) -> Result<Vec<UnsignedMessage>, Error> {
+        let mut umsg: Vec<UnsignedMessage> = Vec::new();
+        let mut msg: Vec<SignedMessage> = Vec::new();
+        for bh in h.blocks().iter() {
+            let (mut bh_umsg_tmp, mut bh_msg_tmp) = self.messages_for_block(bh)?;
+            let bh_umsg = bh_umsg_tmp.as_mut();
+            let bh_msg = bh_msg_tmp.as_mut();
+            umsg.append(bh_umsg);
+            msg.append(bh_msg);
+        }
+        for msg in &msg {
+            umsg.push(msg.message().clone());
```

Comment:
The clone is required because message() returns a reference to the unsigned message and the message field is private so there was no direct way to get the unsigned message without cloning.

Reply:
ah right, well optionally can add a function like: 
```
    pub fn into_message(self) -> UnsignedMessage {
        self.message
    }
```
to avoid this